subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiClassExtractor.kt

```diff
@@ -0,0 +1,176 @@
+import kotlinx.metadata.jvm.signature
+import org.gradle.internal.normalization.java.ApiClassExtractor
+import org.gradle.internal.normalization.java.impl.AnnotationMember
+import org.gradle.internal.normalization.java.impl.ApiMemberWriter
+import org.gradle.internal.normalization.java.impl.ArrayAnnotationValue
+import org.gradle.internal.normalization.java.impl.ClassMember
+import org.gradle.internal.normalization.java.impl.FieldMember
+import org.gradle.internal.normalization.java.impl.InnerClassMember
+import org.gradle.internal.normalization.java.impl.MethodMember
+import org.gradle.internal.normalization.java.impl.MethodStubbingApiMemberAdapter
+import org.gradle.internal.normalization.java.impl.SimpleAnnotationValue
+import org.objectweb.asm.ClassReader
+import org.objectweb.asm.ClassVisitor
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.MethodVisitor
+import org.objectweb.asm.Opcodes
+
+
+class KotlinApiClassExtractor : ApiClassExtractor(
+    emptySet(),
+    { classReader, classWriter ->
+        KotlinApiMemberWriter(
+            MethodStubbingApiMemberAdapter(classWriter),
+            MethodCopyingApiMemberAdapter(classReader, classWriter)
+        )
+    }
+)
+
+
+private
+class KotlinApiMemberWriter(apiMemberAdapter: ClassVisitor, val inlineMethodWriter: MethodCopyingApiMemberAdapter) : ApiMemberWriter(apiMemberAdapter) {
+
+    val kotlinMetadataAnnotationSignature = "Lkotlin/Metadata;"
+
+    val inlineFunctions: MutableSet<String> = HashSet()
+    val internalFunctions: MutableSet<String> = HashSet()
+
+    override fun writeClass(classMember: ClassMember, methods: Set<MethodMember>, fields: Set<FieldMember>, innerClasses: Set<InnerClassMember>) {
+        classMember.annotations.firstOrNull {
+            it.name == kotlinMetadataAnnotationSignature
+        }?.let {
+            when (val kotlinMetadata = KotlinClassMetadata.read(parseKotlinClassHeader(it))) {
+                is KotlinClassMetadata.Class -> kotlinMetadata.toKmClass().extractFunctionMetadata()
+                is KotlinClassMetadata.FileFacade -> kotlinMetadata.toKmPackage().extractFunctionMetadata()
+                else -> {
+                    // KotlinClassMetadata.SyntheticClass || KotlinClassMetadata.MultiFileClassFacade || KotlinClassMetadata.MultiFileClassPart || KotlinClassMetadata.Unknown
+                }
+            }
+        }
+
+        super.writeClass(classMember, methods, fields, innerClasses)
+    }
+
+    override fun writeClassAnnotations(annotationMembers: Set<AnnotationMember>) {
+        super.writeClassAnnotations(annotationMembers.filter { it.name != kotlinMetadataAnnotationSignature }.toSet())
+    }
+
+    override fun writeMethod(method: MethodMember) {
+        if (method.isInternal()) {
+            return
+        }
+
+        if (method.isInline()) {
+            inlineMethodWriter.writeMethod(method)
+        } else {
+            super.writeMethod(method)
+        }
```

Comment:
💄 
```suggestion
        when {
            method.isInternal() -> return
            method.isInline() -> inlineMethodWriter.writeMethod(method)
            else -> super.writeMethod(method)
        }
```

Reply:
Handling this one here https://github.com/gradle/gradle/pull/14821